File: smarthomelibary/DataLibrary/datalibrary/src/main/java/smarthome/datalibrary/database/store/firebase/FirebaseRPInfoStorage.java
package smarthome.datalibrary.database.store.firebase;

import android.support.annotation.NonNull;
import android.util.Log;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.*;
import smarthome.datalibrary.database.store.RPInfoStorage;
import smarthome.datalibrary.database.store.listeners.RPInfoListener;

import static smarthome.datalibrary.database.constants.Constants.FIREBASE_READ_VALUE_ERROR;
import static smarthome.datalibrary.database.constants.Constants.RP_IP_REF;
import static smarthome.datalibrary.BuildConfig.DEBUG;

public class FirebaseRPInfoStorage implements RPInfoStorage {

    private static FirebaseRPInfoStorage instance;

    private DatabaseReference ipRef;


    public static FirebaseRPInfoStorage getInstance() {
        if(instance == null)
            instance = instantiate();

        return instance;
    }

    private static FirebaseRPInfoStorage instantiate() {
        FirebaseAuth auth = FirebaseAuth.getInstance();

        if(auth.getCurrentUser() == null)
            return null;

        else return new FirebaseRPInfoStorage(auth.getCurrentUser().getUid());
    }

    private FirebaseRPInfoStorage(String uid) {
        FirebaseDatabase database = FirebaseDatabase.getInstance();

        ipRef = database.getReference()
                .child(uid)
                .child(RP_IP_REF);

    }

    @Override
    public void postRaspberryIp(String ip) {
        ipRef.setValue(ip);
    }

    @Override
    public void postRaspberryIp(String ip, DatabaseReference.CompletionListener listener) {
        ipRef.setValue(ip, listener);
    }

    @Override
    public void getRaspberryInfo(final RPInfoListener listener) {
        ipRef.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                listener.onRaspberryIpReceived(dataSnapshot.getValue(String.class));
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
                if(DEBUG) Log.d(instance.getClass().getName(), FIREBASE_READ_VALUE_ERROR, databaseError.toException());
            }
        });
    }
}

File: smarthomelibary/DataLibrary/datalibrary/src/main/java/smarthome/datalibrary/database/store/firebase/FirebaseRPInfoStorage.kt
package smarthome.datalibrary.database.store.firebase

import android.util.Log
import com.google.android.gms.tasks.OnFailureListener
import com.google.android.gms.tasks.OnSuccessListener
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.FirebaseFirestore
import smarthome.datalibrary.database.constants.Constants.FIREBASE_READ_VALUE_ERROR
import smarthome.datalibrary.database.constants.Constants.RP_INFO
import smarthome.datalibrary.database.constants.Constants.RP_IP_REF
import smarthome.datalibrary.database.constants.Constants.RP_PORT_REF
import smarthome.datalibrary.database.store.RPInfoStorage
import smarthome.datalibrary.database.store.listeners.RPInfoListener

class FirebaseRPInfoStorage private constructor(
    private val uid: String,
    private val db: FirebaseFirestore = FirebaseFirestore.getInstance()
) : RPInfoStorage {

    private val ref: DocumentReference = db.collection(uid).document(RP_INFO)

    private var rpInfoMap = HashMap<String, Any>()

    override fun postRaspberryIp(
        ip: String,
        successListener: OnSuccessListener<Void>,
        failureListener: OnFailureListener
    ) {
        rpInfoMap[RP_IP_REF] = ip
        ref.set(rpInfoMap)
            .addOnSuccessListener(successListener)
            .addOnFailureListener(failureListener)
    }

    override fun postRaspberryPort(
        port: String,
        successListener: OnSuccessListener<Void>,
        failureListener: OnFailureListener
    ) {
        rpInfoMap[RP_PORT_REF] = port
        ref.set(rpInfoMap)
            .addOnSuccessListener(successListener)
            .addOnFailureListener(failureListener)
    }

    override fun getRaspberryInfo(listener: RPInfoListener) {
        ref.get()
            .addOnSuccessListener { res ->
                run {
                    if (res != null) {
                        val responseInfoMap = res.data
                        listener.onRaspberryIpReceived(responseInfoMap?.get(RP_IP_REF) as String)
                        listener.onRaspberryPortReceived(responseInfoMap[RP_PORT_REF] as String)
                    }
                }
            }
            .addOnFailureListener { exception -> Log.d(javaClass.name, FIREBASE_READ_VALUE_ERROR, exception) }
    }

    companion object {

        private var instance: FirebaseRPInfoStorage? = null


        fun getInstance(): FirebaseRPInfoStorage? {
            if (instance == null)
                instance = instantiate()

            return instance
        }

        private fun instantiate(): FirebaseRPInfoStorage? {
            val auth = FirebaseAuth.getInstance()

            auth.currentUser?.let { return FirebaseRPInfoStorage(it.uid) } ?: return null
        }
    }
}
